ZOOKEEPER-2845: Apply commit log when restarting server.

revans2 opened the pull request:
I will be creating a patch/pull request for 3.4 and 3.5 too, but I wanted to get a pull request up for others to look at ASAP.

I have a version of this based off of #310 at https://github.com/revans2/zookeeper/tree/ZOOKEEPER-2845-orig-test-patch but the test itself is flaky.  Frequently leader election does not go as planned on the test and it ends up failing but not because it ended up in an inconsistent state.

I am happy to answer any questions anyone has about the patch.

mfenes commented on src/java/main/org/apache/zookeeper/server/quorum/LearnerHandler.java:
```diff
@@ -758,6 +760,11 @@ public boolean syncFollower(long peerLastZxid, ZKDatabase db, Leader leader) {
+            } else if (peerLastEpoch != lastProcessedEpoch && !db.isInCommittedLog(peerLastZxid)) {
+                //Be sure we do a snap, because if the epochs are not the same we don't know what
+                // could have happened in between and it may take a TRUNC + UPDATES to get them in SYNC
+                LOG.debug("Will send SNAP to peer sid: {} epochs are too our of sync local 0x{} remote 0x{}",
```
I think there is a typo here: "our of sync"

mfenes commented on src/java/main/org/apache/zookeeper/server/quorum/LearnerHandler.java:
```diff
@@ -758,6 +760,11 @@ public boolean syncFollower(long peerLastZxid, ZKDatabase db, Leader leader) {
                 currentZxid = maxCommittedLog;
                 needOpPacket = false;
                 needSnap = false;
+            } else if (peerLastEpoch != lastProcessedEpoch && !db.isInCommittedLog(peerLastZxid)) {
```
Could you please add a description to the comments above (to "Here are the cases that we want to handle") what this else if case is doing?

revans2 commented:
Thank you to everyone who reviewed the patch, but with the help of Fangmin Lv I found one case that the original patch didn't cover.  I have reworked the patch to cover that case, but to do so I had to take a completely different approach.

I think this is a better approach because it reuses a lot of the code that was originally run to load the database from disk.  So now instead of reloading the entire database from disk, we apply all of the uncommitted transactions in the log to the in memory database.  This should put it in exactly the same state as if we had cleared the data and reloaded it from disk, but with much less overhead.

anmolnar commented:
@revans2 Your latest change looks good to me and a bit safer than the previous one. Would you please consider adding some unit tests to validate the functionality?
What do you think of porting testTxnAheadSnapInRetainDB() test from your codebase?
Maybe I can help making it not flaky, if you think it correctly verifies the original issue.

anmolnar commented on src/java/main/org/apache/zookeeper/server/ZKDatabase.java:
```diff
@@ -233,14 +233,32 @@ public long getDataTreeLastProcessedZxid() {
+     * @throws IOException
+     */
+    public long fastForwardDataBase() throws IOException {
+        PlayBackListener listener = new PlayBackListener(){
```
I think it'd be nice to extract the common logic of these two methods into a operate one.

revans2 replied:
Will do

revans2 commented:
@anmolnar I will add some kind of a test.  I ran into a lot of issues with `testTxnAheadSnapInRetainDB`.  I could not get it to run correctly against master as it would always end up electing the original leader again and the test would fail, but not because it had reproduced the issue.  I finally just did development work based off of the [original patch](https://github.com/apache/zookeeper/compare/master...revans2:ZOOKEEPER-2845-updated-fix?expand=1) and verified that `testTxnAheadSnapInRetainDB` passed, or that if it failed it did so because of leader election.

anmolnar commented:
@revans2 Take a look at `testElectionFraud()` in the same file. Maybe I'm wrong, but it seems to me trying to achieve something similar.

afine commented on src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java:
```diff
@@ -435,7 +435,7 @@ private void waitForOne(ZooKeeper zk, States state) throws InterruptedException
         while (zk.getState() != state) {
             if (iterations-- == 0) {
-                throw new RuntimeException("Waiting too long");
+                throw new RuntimeException("Waiting too long " + zk.getState() + " != " + state);
```
nit: let's minimize unrelated test changes and whitespace changes

anmolnar replied:
Although I agree with you in general, I think this one here is a good addition to test output.

revans2 replied:
@anmolnar  and @afine I put this in for my own debugging and I forgot to remove it.  If you want me to I am happy to either remove it or file a separate JIRA and put it up as a separate pull request, or just leave it.  Either way is fine with me.

afine replied:
Since @anmolnar thinks it is valuable, I think it is fine for it to be left in.

afine commented on src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java:
```diff
@@ -888,4 +888,127 @@ public void testWithOnlyMinSessionTimeout() throws Exception {
     }
 
+    @Test
+    public void testTxnAheadSnapInRetainDB() throws Exception {
```
nit: I don't think we use the terminology "RetainDB" anywhere else. Perhaps we can get rid of "retain"?

revans2 replied:
done

afine commented on src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java:
```diff
@@ -888,4 +888,127 @@ public void testWithOnlyMinSessionTimeout() throws Exception {
+
+        // 6. wait the leader to quit due to no enough followers
+        Thread.sleep(4000);
+        //waitForOne(zk[leader], States.CONNECTING);
```
remove this

revans2 replied:
done

afine commented on src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java:
```diff
@@ -888,4 +888,127 @@ public void testWithOnlyMinSessionTimeout() throws Exception {
+        MainThread mt[] = new MainThread[SERVER_COUNT];
+        ZooKeeper zk[] = new ZooKeeper[SERVER_COUNT];
+        for (int i = 0; i < SERVER_COUNT; i++) {
+            mt[i] = new MainThread(i, clientPorts[i], quorumCfgSection);
```
is there any reason we can't use the existing test infra to clean this up a little

anmolnar replied:
+1
As mentioned testElectionFraud() is a good example for that.

revans2 replied:
I am not super familiar with the test infrastructure.  If you have a suggestion I would love it, otherwise I will look around and see what I can come up with.

anmolnar replied:
Use `LaunchServers(numServers, tickTime)` method in this class.
It gives you a collection of `MainThread` and `ZooKeeper` objects properly initialized.
Test `tearDown()` will care about destroying it.

afine commented on src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java:
```diff
@@ -888,4 +888,127 @@ public void testWithOnlyMinSessionTimeout() throws Exception {
+        // just make sure that we actually did get it in process at the
+        // leader
+        Assert.assertTrue(outstanding.size() == 1);
+        Proposal p = (Proposal) outstanding.values().iterator().next();
```
Do we need this cast?

revans2 replied:
removed the cast

afine commented on src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java:
```diff
@@ -888,4 +888,127 @@ public void testWithOnlyMinSessionTimeout() throws Exception {
+        Assert.assertTrue(p.request.getHdr().getType() == OpCode.create);
+
+        // make sure it has a chance to write it to disk
+        Thread.sleep(1000);
```
There is a lot of `Thread.sleep()` going on and I would like to find a way to minimize that. Apache infra can occasionally be quite slow (it can starve threads) and tests with many `Thread.sleep()`s in them have historically been quite flaky.

So, to the extent that it is possible. I would like to minimize occurrences of `Thread.sleep()`, or at least those outside the context of retry logic.

So perhaps, we can throw `p.qvAcksetPairs.get(0).getAckset().contains(leader);` in a loop waiting one second between iterations.

w.r.t. step 6, we can wait for the leader to enter the looking state.

What do you think?

revans2 replied:
I will see if I can make it work.  I agree I would love to kill as many of the sleeps as possible.

revans2 replied:
I was able to do what you said and drop the 1 second sleep, but the sleep at step 6 I am going to need something else because the leader is only in the looking state for 2 ms.  Leader election happens way too fast for us to be able to catch that by polling.  

If I remove the 4 second sleep it does not trigger the error case, I don't completely know why.  I'll spend some time looking at it, but if you have any suggestions I would appreciate it.

afine replied:
@revans2 take a look at `testElectionFraud`, specifically: https://github.com/apache/zookeeper/blob/master/src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java#L383 and https://github.com/apache/zookeeper/blob/master/src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java#L397

You can manually start and stop leader election, I think that may solve this problem.

revans2 commented:
@afine and @anmolnar I think I have addressed all of your review comments, except for the one about the change to `waitForOne` and I am happy to adjust however you want there.

afine commented on src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java:
```diff
@@ -465,6 +470,37 @@ private void waitForAll(ZooKeeper[] zks, States state) throws InterruptedExcepti
+            }
+        }
+
+        public void restartClient(int i, Watcher watcher) throws IOException {
```
annoying nitpick: let's use a better argument name than `i`

afine commented on src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java:
```diff
@@ -465,6 +470,37 @@ private void waitForAll(ZooKeeper[] zks, States state) throws InterruptedExcepti
+            zk[i] = new ZooKeeper("127.0.0.1:" + clientPorts[i], ClientBase.CONNECTION_TIMEOUT, watcher);
+        }
+
+        public int findLeader() {
```
there are other places in this test class that benefit from this refactoring. Would you mind cleaning that up?

afine commented on src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java:
```diff
@@ -888,4 +923,103 @@ public void testWithOnlyMinSessionTimeout() throws Exception {
+        sleepTime = 0;
+        Follower f = servers.mt[leader].main.quorumPeer.follower;
+        while (f == null || !f.isRunning()) {
+            if (sleepTime > 10_000) {
```
nitpick: can we reuse the ticktime here to make the relationship more obvious?

afine commented on src/java/test/org/apache/zookeeper/server/quorum/QuorumPeerMainTest.java:
```diff
@@ -888,4 +923,103 @@ public void testWithOnlyMinSessionTimeout() throws Exception {
+            sleepTime += 100;
+            f = servers.mt[leader].main.quorumPeer.follower;
+        }
+        servers.mt[leader].shutdown();
```
why do we need this?

revans2 replied:
It is a lot of very specific steps that make the data inconsistency show up.  This is needed to force the transaction log to be replayed which has an edit in it that wasn't considered as a part of leader election.

revans2 commented:
@afine 

I have addressed you most recent comments.  If you want me to squash commits please let me know.

I have a pull request for the 3.5 branch #454 and for the 3.4 branch #455.  I will be spending some time porting the test to them, and let you know when it is ready.

revans2 commented:
@afine all of the changes in this branch are now in the pull requests to the 3.5 and 3.5 branches,

afine commented:
Thanks @revans2. I merged this and the PR's for 3.4 and 3.5

revans2 commented:
Thanks @afine I closed them.
